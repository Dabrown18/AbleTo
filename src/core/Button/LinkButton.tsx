import React, { FunctionComponent, useCallback } from 'react';
import { Text, TextProps } from 'react-native';
import styled from 'styled-components/native';

import { openLink } from '@src/services/helpers';

type Props = {
    title: string;
    link?: string;
} & TextProps;

const LinkButton: FunctionComponent<Props> = ({ title, link, ...rest }) => {
    // Normal usage entails no passed children.
    // If any are present though, we override them with
    // the 'title' prop, which is the text we want to display
    const spread = { ...rest, children: title };

    const handlePress = useCallback(() => openLink(link), [link]);

    return <StyledBodyText accessibilityLabel={title} accessibilityRole="link" onPress={handlePress} {...spread} />;
};

const StyledBodyText = styled(Text)`
    font-family: ${({ theme }) => theme.fonts.FAKT_PRO_NORMAL};
    color: ${({ theme }) => theme.colors.teal600};

    font-weight: 400;
    font-size: 16px;
    line-height: 24px;

    text-decoration-line: underline;
`;

export default LinkButton;
