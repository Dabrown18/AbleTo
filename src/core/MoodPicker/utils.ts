import { TAU } from 'react-native-redash';

import colors from '../colors';

export const PICKER_SIZE = 220;
export const STROKE = 20;
export const KNOB_SIZE = 30;
export const BORDER_SIZE = 6;
export const R = (PICKER_SIZE - STROKE) / 2;
export const CENTER = { x: R, y: R };
const BUFFER = 0.17;

export type MoodRating = typeof MOOD_RATINGS[0];

export const normalize = (value: number) => {
    'worklet';

    const mod = value % TAU;
    return mod < 0 ? TAU + mod : mod;
};

export const MOOD_RATINGS = [
    { color: colors.teal300, valueInt: 7 },
    { color: '#B3CFBC', valueInt: 6 },
    { color: '#C8D3B3', valueInt: 5 },
    { color: colors.yellow200, valueInt: 4 },
    { color: '#F4C491', valueInt: 3 },
    { color: '#F2B487', valueInt: 2 },
    { color: colors.orange300, valueInt: 1 },
];

export const getCurrentMood = (rad: number) => {
    'worklet';

    // Adds 10 degrees (or 0.17 radians)
    // to offset the size of the rotator knob
    const radian = rad + BUFFER;

    if (radian >= 0 && radian < 0.89) {
        return MOOD_RATINGS[1];
    } else if (radian >= 0.89 && radian < 1.79) {
        return MOOD_RATINGS[0];
    } else if (radian >= 1.79 && radian < 2.69) {
        return MOOD_RATINGS[6];
    } else if (radian >= 2.69 && radian < 3.58) {
        return MOOD_RATINGS[5];
    } else if (radian >= 3.58 && radian < 4.48) {
        return MOOD_RATINGS[4];
    } else if (radian >= 4.48 && radian < 5.38) {
        return MOOD_RATINGS[3];
    } else {
        return MOOD_RATINGS[2];
    }
};

export const getRadianForCurrentMood = (moodValue: number) => {
    'worklet';
    switch (moodValue) {
        case 7:
            return 0.89 + BUFFER;
        case 6:
            return 0 + BUFFER;
        case 5:
            return 5.4 + BUFFER;
        case 4:
            return 4.48 + BUFFER;
        case 3:
            return 3.58 + BUFFER;
        case 2:
            return 2.69 + BUFFER;
        case 1:
            return 1.79 + BUFFER;
        default:
            return 0 + BUFFER;
    }
};

// Removes 1/Nth of TAU (the full circle rotation) from the current radian
// Where N is the number of mood ratings we have
export const getNextMood = (rad: number) => normalize(rad - TAU / MOOD_RATINGS.length);
