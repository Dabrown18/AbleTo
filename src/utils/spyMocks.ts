import { AccessibilityInfo } from 'react-native';

import { mockAddEventListener } from './mocks/accessibilityInfo';

// Add mocks that we'd like to spy on globally in any tests.
// This file is referenced in `jest.config.js`.
jest.spyOn(AccessibilityInfo, 'addEventListener').mockImplementation(mockAddEventListener as any);
