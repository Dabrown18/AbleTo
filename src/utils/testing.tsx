import { NavigationContainer } from '@react-navigation/native';
import { render } from '@testing-library/react-native';
import React, { ReactNode } from 'react';
import { Provider as PaperProvider } from 'react-native-paper';
import { QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import TestRenderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

import server from './mocks/server';
import { rootNavigation } from '@src/activities/controller/ActivityController';
import linking from '@src/config/deeplinks';
import queryClient from '@src/config/queryClient';
import colors from '@src/core/colors';
import fonts from '@src/core/fonts';
import LoadingIndicator from '@src/core/LoadingIndicator/LoadingIndicator';
import store from '@src/redux/store';

const theme = {
    colors,
    fonts,
};

// Globally exposed testing utils that can be reused across multiple files

export const renderWithTheme = (children: ReactNode) => render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

// Currently used in tests that involve controller initialization
// as the controller uses the navigator ref
export const renderWithNavigator = (children: ReactNode) => {
    return render(<NavigationContainer ref={rootNavigation}>{children}</NavigationContainer>);
};

const applyProviders = (children: ReactNode) => {
    return (
        <Provider store={store}>
            <QueryClientProvider client={queryClient}>
                <ThemeProvider theme={theme}>
                    <NavigationContainer linking={linking} ref={rootNavigation} fallback={<LoadingIndicator />}>
                        <PaperProvider>{children}</PaperProvider>
                    </NavigationContainer>
                </ThemeProvider>
            </QueryClientProvider>
        </Provider>
    );
};

export const renderWithAppProviders = (children: ReactNode) => render(applyProviders(children));

export const createWithAppProviders = (children: ReactNode) => TestRenderer.create(applyProviders(children));

/**
 * Sets up a request interception layer.
 * Should be called when tests involve API calls.
 * @param clearCache Clear QueryClient cache. Defaults to true.
 */
export const initServer = (clearCache = true) => {
    // Sets up the requests interception layer before all tests.
    beforeAll(() => server.listen());
    // Removes any request handlers that were added on runtime
    // and resets QueryClient as consecutive requests to the
    // same endpoint get cached.
    afterEach(() => {
        server.resetHandlers();
        if (clearCache) queryClient.clear();
    });
    // Clean up after all tests are done, preventing this
    // interception layer from affecting irrelevant tests.
    afterAll(() => server.close());
};
