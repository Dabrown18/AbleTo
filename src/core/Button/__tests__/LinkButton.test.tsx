import React from 'react';

import LinkButton from '../LinkButton';
import { renderWithTheme } from '@src/utils/testing';

test('LinkButton snapshot', () => {
    const tree = renderWithTheme(<LinkButton title="Click me!" />).toJSON();

    expect(tree).toMatchSnapshot();
});
