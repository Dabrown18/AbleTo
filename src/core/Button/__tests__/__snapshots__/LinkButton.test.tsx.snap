// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkButton snapshot 1`] = `
<Text
  accessibilityLabel="Click me!"
  accessibilityRole="link"
  onPress={[Function]}
  style={
    Array [
      Object {
        "color": "#2B7D82",
        "fontFamily": "fakt_pro_normal",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 24,
        "textDecorationLine": "underline",
      },
    ]
  }
>
  Click me!
</Text>
`;
